apiml-security-common/src/main/java/org/zowe/apiml/security/common/error/AuthExceptionHandler.java

```diff
@@ -38,143 +40,190 @@
+    private interface ExceptionHandler {
+        void handle(RuntimeException ex, HandlerContext ctx);
+    }
+
+    private final Map<Class<? extends RuntimeException>, ExceptionHandler> exceptionHandlers = Map.ofEntries(
+        entry(InsufficientAuthenticationException.class,
+            (ex, ctx) -> handleAuthenticationRequired(ctx.requestUri, ctx.function, ctx.addHeader, ex)),
+        entry(BadCredentialsException.class,
+            (ex, ctx) -> handleBadCredentials(ctx.requestUri, ctx.function, ex)),
+        entry(AuthenticationCredentialsNotFoundException.class,
+            (ex, ctx) -> handleAuthenticationCredentialsNotFound(ctx.requestUri, ctx.function, ex)),
+        entry(AuthMethodNotSupportedException.class,
+            (ex, ctx) -> handleAuthMethodNotSupported(ctx.requestUri, ctx.function, ex)),
+        entry(TokenNotValidException.class,
+            (ex, ctx) -> handleTokenNotValid(ctx.requestUri, ctx.function, ctx.addHeader, ex)),
+        entry(NoMainframeIdentityException.class,
+            (ex, ctx) -> handleNoMainframeIdentity(ctx.requestUri, ctx.function, ctx.addHeader, (NoMainframeIdentityException) ex)),
+        entry(TokenNotProvidedException.class,
+            (ex, ctx) -> handleTokenNotProvided(ctx.requestUri, ctx.function, ex)),
+        entry(TokenExpireException.class,
+            (ex, ctx) -> handleTokenExpire(ctx.requestUri, ctx.function, ex)),
+        entry(TokenFormatNotValidException.class,
+            (ex, ctx) -> handleTokenFormatException(ctx.requestUri, ctx.function, ex)),
+        entry(AccessTokenBodyNotValidException.class,
+            (ex, ctx) -> handleInvalidAccessTokenBodyException(ctx.requestUri, ctx.function, ex)),
+        entry(InvalidCertificateException.class,
+            (ex, ctx) -> handleInvalidCertificate(ctx.function, ex)),
+        entry(ZosAuthenticationException.class,
+            (ex, ctx) -> handleZosAuthenticationException(ctx.function, (ZosAuthenticationException) ex)),
+        entry(InvalidTokenTypeException.class,
+            (ex, ctx) -> handleInvalidTokenTypeException(ctx.requestUri, ctx.function, ex)),
+        entry(AuthenticationServiceException.class,
+            (ex, ctx) -> handleAuthenticationException(ctx.requestUri, ctx.function, ex)),
+        entry(AuthenticationException.class,
+            (ex, ctx) -> handleAuthenticationException(ctx.requestUri, ctx.function, ex)),
+        entry(ServiceNotAccessibleException.class,
+            (ex, ctx) -> handleServiceNotAccessibleException(ctx.requestUri, ctx.function, ex))
+    );
+
+    private ExceptionHandler resolveHandler(RuntimeException ex) {
+        Class<?> exClass = ex.getClass();
+        while (exClass != null && RuntimeException.class.isAssignableFrom(exClass)) {
+            ExceptionHandler handler = exceptionHandlers.get(exClass);
+            if (handler != null) {
+                return handler;
+            }
+            exClass = exClass.getSuperclass();
+        }
```

Comment:
```suggestion
        while (exClass != null) {
            ExceptionHandler handler = exceptionHandlers.get(exClass);
            if (handler != null) {
                return handler;
            }
            if (exClass == exClass.getSuperclass()) {
                return null;
            }
            exClass = exClass.getSuperclass();
        }
```

It is a rare case, but sometimes the cause could be the same as the exception (to avoid an infinite loop). And I guess checking RuntimeException does not worth it, the check for null is enough.

Reply:
`RuntimeException.class.isAssignableFrom(exClass)` should still be worth it, as we might still end up with matching Exception, Throwable or Object by mistake I believe.
exClass.getSuperclass() == exClass I think should never happen, but I can keep it for safety